pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractReplicator.java

```diff
@@ -83,6 +83,7 @@ public AbstractReplicator(String topicName, String replicatorPrefix, String loca
                 .enableBatching(false)
                 .sendTimeout(0, TimeUnit.SECONDS) //
                 .maxPendingMessages(producerQueueSize) //
+                .blockIfQueueFull(true)
```

Comment:
This would mean blocking the IO threads (or other shared threads) in the broker causing a deadlock

Reply:
We need to find a better solution to have a broker-wide flow control mechanism for replicators that can be tied to the client memory limit.

Reply:
Thanks for reviewing @merlimat . Yes, blocking would be problematic in this case.

Would you mind also reviewing #15723 since the memory limit is currently causing regressions in geo-replication in 2.10 .